src/Dictionaries/HierarchyDictionariesUtils.h

```diff
@@ -0,0 +1,468 @@
+
+    template <typename T>
+    struct IsKeyValidFuncInterface
+    {
+        bool operator()(T key [[maybe_unused]]) { return false; }
+    };
+
+    template <typename T>
+    struct GetParentKeyFuncInterface
+    {
+        std::optional<T> operator()(T key [[maybe_unused]]) { return {}; }
+    };
+
+    /** Calculate hierarchy for keys iterating the hierarchy from child to parent using get_parent_key_func provided by client.
+      * Hierarchy iteration is stopped if key equals null value, get_parent_key_func returns null optional, or hierarchy depth
+      * greater or equal than DBMS_HIERARCHICAL_DICTIONARY_MAX_DEPTH.
+      * IsKeyValidFunc used for each input hierarchy key, if it returns false result hierarchy for that key will have size 0.
+      * Hierarchy result is ElementsAndOffsets structure, for each element there is hierarchy array,
+      * with size offset[element_index] - (element_index > 0 ? offset[element_index - 1] : 0).
+      *
+      * Example:
+      * id  parent_id
+      * 1   0
+      * 2   1
+      * 3   1
+      * 4   2
+      *
+      * If hierarchy_null_value will be 0. Requested keys [1, 2, 3, 4, 5].
+      * Result: [1], [2, 1], [3, 1], [4, 2, 1], []
+      * Elements: [1, 2, 1, 3, 1, 4, 2, 1]
+      * Offsets: [1, 3, 5, 8, 8]
+      */
+    template <typename KeyType, typename IsKeyValidFunc, typename GetParentKeyFunc>
+    ElementsAndOffsets<KeyType> getHierarchy(
+        const PaddedPODArray<KeyType> & keys,
+        const KeyType & hierarchy_null_value,
+        IsKeyValidFunc && is_key_valid_func,
+        GetParentKeyFunc && get_parent_key_func)
+    {
+        size_t hierarchy_keys_size = keys.size();
+
+        PaddedPODArray<KeyType> elements;
+        elements.reserve(hierarchy_keys_size);
+
+        PaddedPODArray<IColumn::Offset> offsets;
+        offsets.reserve(hierarchy_keys_size);
+
+        struct OffsetInArray
+        {
+            size_t offset_index;
+            size_t array_element_offset;
+        };
+
+        HashMap<KeyType, OffsetInArray> already_processes_keys_to_offset;
+        already_processes_keys_to_offset.reserve(hierarchy_keys_size);
+
+        for (size_t i = 0; i < hierarchy_keys_size; ++i)
+        {
+            auto hierarchy_key = keys[i];
+            size_t current_hierarchy_depth = 0;
+
+            bool is_key_valid = std::forward<IsKeyValidFunc>(is_key_valid_func)(hierarchy_key);
+
+            if (!is_key_valid)
+            {
+                offsets.emplace_back(elements.size());
+                continue;
+            }
+
+            while (true)
+            {
+                const auto * it = already_processes_keys_to_offset.find(hierarchy_key);
+
+                if (it)
+                {
+                    const auto & index = it->getMapped();
+
+                    size_t offset = index.offset_index;
+
+                    bool is_loop = (offset == offsets.size());
+
+                    if (unlikely(is_loop))
+                        break;
```

Comment:
Maybe better throw exception to tell client about a loop?

Reply:
Previous imlementation does not throw exception if there is loop in hierarchy. We cannot do it too because of backward compatibility. Check test 01251_dict_is_in_infinite_loop.sql.